src/coprocessor/dag/rpn_expr/types/expr_eval.rs

```diff
@@ -0,0 +1,1397 @@
+        &self,
+        context: &mut EvalContext,
+        rows: usize,
+        schema: &[FieldType],
+        columns: &mut LazyBatchColumnVec,
+        outputs: &mut [bool], // modify an existing buffer to avoid repeated allocation
+    ) -> Result<()> {
+        use crate::coprocessor::codec::data_type::AsMySQLBool;
+
+        assert!(outputs.len() >= rows);
+        let values = self.eval(context, rows, schema, columns)?;
+        match values {
+            RpnStackNode::Scalar { value, .. } => {
+                let b = value.as_mysql_bool(context)?;
+                for i in 0..rows {
+                    outputs[i] = b;
+                }
+            }
+            RpnStackNode::Vector { value, .. } => {
+                let vec_ref = value.as_ref();
+                assert_eq!(vec_ref.len(), rows);
+                vec_ref.eval_as_mysql_bools(context, outputs)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    #![allow(clippy::float_cmp)]
+
+    use super::*;
+
+    use cop_datatype::{EvalType, FieldTypeAccessor, FieldTypeTp};
+    use tipb::expression::FieldType;
+
+    use super::super::RpnFnCallPayload;
+
+    use crate::coprocessor::codec::batch::LazyBatchColumn;
+    use crate::coprocessor::dag::expr::EvalContext;
+    use crate::coprocessor::dag::rpn_expr::RpnExpressionBuilder;
+    use crate::coprocessor::Result;
+
+    /// Single constant node
+    #[test]
+    fn test_eval_single_constant_node() {
+        let rpn_nodes = vec![RpnExpressionNode::Constant {
+            value: ScalarValue::Real(Some(1.5)),
+            field_type: {
+                let mut ft = FieldType::new();
+                ft.as_mut_accessor().set_tp(FieldTypeTp::Double);
+                ft
+            },
+        }];
+        let exp = RpnExpression::from(rpn_nodes);
+        let mut ctx = EvalContext::default();
+        let mut columns = LazyBatchColumnVec::from(vec![]);
+        let result = exp.eval(&mut ctx, 10, &[], &mut columns);
+        let val = result.unwrap();
+        assert!(val.is_scalar());
+        assert_eq!(*val.scalar_value().unwrap().as_real(), Some(1.5));
+        assert_eq!(val.field_type().tp(), FieldTypeTp::Double);
+    }
+
+    /// Single column node
+    #[test]
+    fn test_eval_single_column_node() {
+        let columns = LazyBatchColumnVec::from(vec![
+            {
+                // this column is not referenced
+                let mut col = LazyBatchColumn::decoded_with_capacity_and_tp(5, EvalType::Real);
+                col.mut_decoded().push_real(Some(1.0));
+                col.mut_decoded().push_real(None);
+                col.mut_decoded().push_real(Some(7.5));
+                col.mut_decoded().push_real(None);
+                col.mut_decoded().push_real(None);
+                col
+            },
+            {
+                let mut col = LazyBatchColumn::decoded_with_capacity_and_tp(5, EvalType::Int);
+                col.mut_decoded().push_int(Some(1));
```

Comment:
Could we use the loop to make the code more clear?

Reply:
What kind of loop are you expecting? Looks like current code is already the most readable one. No branches, no checks, no states.

Reply:
How about
```
    for v1,v2 in data{
             col1.mut_decoded().push_real(v1);
             col2.mut_decoded().push_int(v2)
     }
```

And when we want to add more rows one day, just append in data?

Reply:
According to my understanding, the real problem is that we don't have a utility function to push multiple values in VectorValue? If we have such function, then we achieve both simplicity and DRY. Let's do it later because we will have one when aggregation function is ready.